gossip/hub/handlers.go

```diff
@@ -0,0 +1,664 @@
+	}
+	identityHash := sha256.Sum256(timelessData)
+
+	// The full leaf has a timestamp and signature (but there may be a pre-existing entry
+	// with different values for these fields).
+	entry.SourceSignature = req.SourceSignature
+	entry.HubTimestamp = timeNanos
+	leafData, err := tls.Marshal(entry)
+	if err != nil {
+		glog.V(1).Infof("%s: failed to tls.Marshal hub leaf for %q: %v", c.hubPrefix, req.SourceID, err)
+		return http.StatusBadRequest, fmt.Errorf("failed to create timestamped hub leaf: %v", err)
+	}
+
+	leaf := trillian.LogLeaf{
+		LeafValue:        leafData,
+		LeafIdentityHash: identityHash[:],
+	}
+
+	// Send the leaf on to the Trillian Log server.
+	glog.V(2).Infof("%s: AddLogHead => grpc.QueueLeaves", c.hubPrefix)
+	rsp, err := c.rpcClient.QueueLeaves(ctx, &trillian.QueueLeavesRequest{LogId: c.logID, Leaves: []*trillian.LogLeaf{&leaf}})
+	glog.V(2).Infof("%s: AddLogHead <= grpc.QueueLeaves err=%v", c.hubPrefix, err)
+	if err != nil {
+		return c.toHTTPStatus(err), fmt.Errorf("backend QueueLeaves request failed: %v", err)
+	}
+	if rsp == nil {
+		return http.StatusInternalServerError, errors.New("missing QueueLeaves response")
+	}
+	if len(rsp.QueuedLeaves) != 1 {
+		return http.StatusInternalServerError, fmt.Errorf("unexpected QueueLeaves response leaf count: %d", len(rsp.QueuedLeaves))
+	}
+	entryData := rsp.QueuedLeaves[0].Leaf.LeafValue
+
+	// Always use the data from returned leaf as the basis for an signed gossip
+	// timestamp -- there may be an earlier logged entry, which may also have a
+	// different signature than the submitted blob (for a non-deterministic
+	// signature scheme like ECDSA).
+	var loggedEntry api.TimestampedEntry
+	if rest, err := tls.Unmarshal(entryData, &loggedEntry); err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to reconstruct TimestampedEntry: %s", err)
+	} else if len(rest) > 0 {
+		return http.StatusInternalServerError, fmt.Errorf("extra data (%d bytes) on reconstructing TimestampedEntry", len(rest))
+	}
+
+	signature, err := c.signer.Sign(entryData)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to sign SGT data: %v", err)
+	}
+	jsonRsp := api.AddSignedBlobResponse{TimestampedEntryData: entryData, HubSignature: signature}
+
+	w.Header().Set(contentTypeHeader, contentTypeJSON)
+	jsonData, err := json.Marshal(&jsonRsp)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to JSON-marshal add-signed-blob rsp: %s", err)
+	}
+
+	if _, err = w.Write(jsonData); err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to write add-signed-blob rsp: %s", err)
+	}
+
+	glog.V(3).Infof("%s: AddSignedBlob <= SGT", c.hubPrefix)
+	return http.StatusOK, nil
+}
+
+// GetLogRoot retrieves a log root for the given Trillian log.
+func GetLogRoot(ctx context.Context, client trillian.TrillianLogClient, logID int64, prefix string) (*types.LogRootV1, error) {
```

Comment:
Is this the same thing as `google/trillian/client/UpdateRoot` ?

Reply:
No -- `client.UpdateRoot` enforces serialization and maintains a persistent copy of the current root, which isn't needed here.